### ISSUE TITLE
Issue with TypeScript and useStoreApi

### ISSUE THREAD

**Issue by fryck** [pmndrs/zustand#626]
Hello !

I have an issue with typescript and the useStoreApi hook:

```
export type CustomStoreApi = StoreApi<IState> & StoreApiWithSubscribeWithSelector<IState>;

const zustandContext = createContext<IState, UseBoundStore<IState, CustomStoreApi>>();

export const StoreProvider = zustandContext.Provider;

export const useStore = zustandContext.useStore;
export const useStoreApi = zustandContext.useStoreApi;
```

When I try to use my hook `useStoreApi` even if I properly typed the type above, I have this error :

![Screenshot 2021-11-02 at 10 54 05](https://user-images.githubusercontent.com/1364225/139824940-b51382ae-9be0-47a3-965a-873796820235.png)

We can see that type is properly set

![Screenshot 2021-11-02 at 10 59 03](https://user-images.githubusercontent.com/1364225/139825750-24a136a5-ed64-4ff8-a387-60eb92ee4e9d.png)

Could you help me please ? thanks !

**Comment by dai-shi** [comment]
Thanks for reporting. We need more feedback on typing and improve the docs.

At a glance, I thought it looks good too. But, there's a pitfall.
Please try this:
```ts
export type CustomStoreApi = StoreApiWithSubscribeWithSelector<IState>;
```

This should work. We have a tests in `tests/context.test.tsx`.
Also, note that this will probably be fixed in v4, I suppose, with #604.
All that said, typing with custom store with middleware is nightmare.
We don't have a silver bullet solution for this, and need to find a way to live with it.

**Comment by fryck** [comment]
> Thanks for reporting. We need more feedback on typing and improve the docs.
> 
> At a glance, I thought it looks good too. But, there's a pitfall. Please try this:
> 
> ```ts
> export type CustomStoreApi = StoreApiWithSubscribeWithSelector<IState>;
> ```
> 
> This should work. We have a tests in `tests/context.test.tsx`. Also, note that this will probably be fixed in v4, I suppose, with #604. All that said, typing with custom store with middleware is nightmare. We don't have a silver bullet solution for this, and need to find a way to live with it.

Yes it works, thanks for your quick replay (again) it's a pleasure !
